components/remote_settings/src/lib.rs

```diff
@@ -0,0 +1,195 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+pub mod error;
+use config::RemoteSettingsConfig;
+pub use error::{RemoteSettingsError, Result};
+mod config;
+use rs_client::{Client, ClientConfig};
+use serde::Deserialize;
+use std::{fs::File, io::prelude::Write};
+
+uniffi_macros::include_scaffolding!("remotesettings");
+
+pub struct RemoteSettings {
+    pub config: RemoteSettingsConfig,
+    client: Client,
+}
+
+impl RemoteSettings {
+    pub fn new(config: RemoteSettingsConfig) -> Result<Self> {
+        Ok(RemoteSettings {
+            config: config.clone(),
+            client: Client::new(config.into())?,
+        })
+    }
+
+    pub fn get_records(&self) -> Result<Vec<JsonObject>> {
+        let resp = self.client.get_records()?;
+        Ok(resp.json::<RemoteSettingsResponse<_>>()?.data)
+    }
+
+    pub fn download_attachment_to_path(
+        &self,
+        attachment_location: String,
+        path: String,
+    ) -> Result<()> {
+        let resp = self.client.get_attachment(&attachment_location)?;
+        let mut file = File::create(path)?;
+        file.write_all(&resp.body)?;
+        Ok(())
+    }
+
+    pub fn get_records_since(&self, last_modified: i64) -> Result<Vec<JsonObject>> {
+        let resp = self.client.get_records_since(last_modified)?;
+        Ok(resp.json::<RemoteSettingsResponse<_>>()?.data)
+    }
+}
+
+impl From<RemoteSettingsConfig> for ClientConfig {
+    fn from(c: RemoteSettingsConfig) -> ClientConfig {
+        ClientConfig {
+            server_url: c.server_url,
+            bucket_name: c.bucket_name,
+            collection_name: c.collection_name,
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct RemoteSettingsResponse<T> {
+    data: Vec<T>,
+}
+
+#[derive(Deserialize)]
+pub struct JsonObject(pub serde_json::Value);
+
+impl UniffiCustomTypeConverter for JsonObject {
+    type Builtin = String;
+
+    fn into_custom(val: Self::Builtin) -> uniffi::Result<Self> {
+        Ok(JsonObject(serde_json::from_str(&val)?))
+    }
+
+    fn from_custom(obj: Self) -> Self::Builtin {
+        serde_json::to_string_pretty(&obj.0).unwrap()
```

Comment:
seems mildly wasteful for this to be pretty, especially given the Kotlin code never actually sees the string (ie, it doesn't even seem able to be inspected anywhere for debugging)

Re the type stuff above, the `.0` here (and below in the test get removed)

Reply:
That's a good point, I don't think my brain had caught up from thinking about the result being a list of strings